core/platform/src/allows/allowed_tests.rs

```diff
@@ -0,0 +1,184 @@
+// variable behave identically to mutations performed by the kernel. This
+// pattern allows us to use `cargo miri test` to not only execute the unit
+// tests, but to test whether Allowed would encounter undefined behavior when
+// interacting with a real Tock kernel.
+//
+// [1] https://plv.mpi-sws.org/rustbelt/stacked-borrows/paper.pdf
+struct KernelPtr<'b, T: 'b> {
+    ptr: NonNull<T>,
+
+    // We need to consume the 'b lifetime. This is very similar to Allowed's
+    // implementation.
+    _phantom: PhantomData<&'b mut T>,
+}
+
+impl<'b, T: 'b> KernelPtr<'b, T> {
+    // The constructor for KernelPtr; simulates allow(). Returns both the
+    // Allowed instance the Platform would return and a KernelPtr the test can
+    // use to simulate a kernel.
+    pub fn allow(buffer: &'b mut T) -> (Allowed<'b, T>, KernelPtr<'b, T>) {
+        let ptr = NonNull::new(buffer).unwrap();
+        // Discard buffer *without* creating a reference to it, as would be done
+        // if we called drop().
+        let _ = buffer;
+        // All 3 preconditions of Allowed::new are satisfied by the fact that
+        // `buffer` is directly derived from a &'b mut T.
+        let allowed = unsafe { Allowed::new(ptr) };
+        let kernel_ptr = KernelPtr {
+            ptr,
+            _phantom: PhantomData,
+        };
+        (allowed, kernel_ptr)
+    }
+
+    // Replaces the value in the buffer with a new one. Does not drop the
+    // existing value.
+    pub fn set(&self, value: T) {
+        unsafe {
+            core::ptr::write(self.ptr.as_ptr(), value);
+        }
+    }
+}
+
+impl<'b, T: Copy + 'b> KernelPtr<'b, T> {
+    // Copies the contained value out of the buffer.
+    pub fn get(&self) -> T {
+        unsafe { core::ptr::read(self.ptr.as_ptr()) }
+    }
+}
+
+impl<'b, 'f: 'b> KernelPtr<'b, DropCheck<'f>> {
+    // Retrieves the value of the contained DropCheck.
+    pub fn value(&self) -> usize {
+        let drop_check = unsafe { core::ptr::read(self.ptr.as_ptr()) };
+        let value = drop_check.value;
+        core::mem::forget(drop_check);
+        value
+    }
+}
+
+#[test]
+fn set() {
+    let dropped1 = core::cell::Cell::new(false);
+    let dropped2 = core::cell::Cell::new(false);
+    let dropped3 = core::cell::Cell::new(false);
+    let mut buffer = DropCheck {
+        flag: Some(&dropped1),
+        value: 1,
+    };
+    let (allowed, kernel_ptr) = KernelPtr::allow(&mut buffer);
+    assert_eq!(kernel_ptr.value(), 1);
+    assert_eq!(dropped1.get(), false);
+
+    // Simulate the kernel replacing the value in buffer. We don't drop the
+    // existing value.
+    kernel_ptr.set(DropCheck {
+        flag: Some(&dropped2),
+        value: 2,
+    });
+    allowed.set(DropCheck {
+        flag: Some(&dropped3),
+        value: 3,
+    });
+    assert_eq!(kernel_ptr.value(), 3);
+    assert_eq!(dropped1.get(), false);
+    assert_eq!(dropped2.get(), false);
+    assert_eq!(dropped3.get(), false);
+}
+
```

Comment:
An additional test I wrote to help me understand how dropping works with `Allowed`. Feel free to include it if you want, or ignore it.

```rust
#[test]
fn drop_test() {
    let dropped1 = core::cell::Cell::new(false);
    let dropped2 = core::cell::Cell::new(false);
    let dropped3 = core::cell::Cell::new(false);
    let dropped4 = core::cell::Cell::new(false);
    let mut buffer = DropCheck {
        flag: Some(&dropped1),
        value: 1,
    };
    let (allowed, kernel_ptr) = KernelPtr::allow(&mut buffer);
    assert_eq!(kernel_ptr.value(), 1);
    assert_eq!(dropped1.get(), false);

    let returned = allowed.replace(DropCheck {
        flag: Some(&dropped2),
        value: 2,
    });
    assert_eq!(returned.value, 1);
    assert_eq!(kernel_ptr.value(), 2);
    assert_eq!(dropped1.get(), false);
    drop(returned); // drops value inserted by userspace, because kernel has not overwritten it.
    assert_eq!(dropped1.get(), true);
    assert_eq!(dropped2.get(), false);
    kernel_ptr.set(DropCheck {
        flag: Some(&dropped3),
        value: 3,
    });
    // drop implicitly
    allowed.replace(DropCheck {
        flag: Some(&dropped4),
        value: 4,
    });
    assert_eq!(dropped2.get(), false); // overwritten by kernel, so not dropped by replace
    assert_eq!(dropped3.get(), true); // inserted by kernel, but taken by userspace so dropped by replace
    assert_eq!(dropped4.get(), false); // still allowed
}
```

Reply:
Thanks for the test! I added a `Copy` bound to `Allowed` so this test is no longer relevant or possible.